docs/config.md
# Config

Configuration is primarily provided to Frigate-Notify via a [Config File](#config-file). A [sample config](#sample-config) file is available for customization at the bottom of this page.

Additionaly, a few [options](#config-options) are available through command line flags and/or environment variables.

## Config Options

The following options are available as command line flags or environment variables:

| Flag     | Environment Variable | Description                             |
|----------|----------------------|-----------------------------------------|
| -c       | FN_CONFIGFILE        | Specify alternate config file location  |
| -debug   | FN_DEBUG             | Set to `true` to enable debug logging   |
| -jsonlog | FN_JSONLOG           | Set to `true` to enable logging in JSON |

## Config File

The following section details options available via the `config.yml` file. Config snippets are provided for each section, however it is recommended to check the full [sample config](#sample-config) below.

### Frigate

#### Server

- **server** (Required)
    - IP, hostname, or URL of the Frigate NVR
    - If IP or hostname specified, app will prepend `http://`
    - If Frigate is not behind a reverse proxy, append port number if necessary
- **ignoressl** (Optional - Default: `false`)
    - Set to `true` to allow self-signed certificates for `server`
- **public_url** (Optional)
    - Should be set if Frigate is available via an external, public URL
    - This value is used for the links used in notifications
    - Format should be full URL (example: `https://nvr.your.public.domain.tld`)
- **headers** (Optional)
    - Send additional HTTP headers to Frigate
    - Useful for things like authentication
    - Header format: `Header: Value`
    - Example: `Authorization: Basic abcd1234`

```yaml title="Config File Snippet"
frigate:
  server: nvr.your.domain.tld
  ignoressl: true
  public_url: https://nvr.your.public.domain.tld
  headers:
    - Authorization: Basic abcd1234
```

#### WebAPI

!!! note
    Only one monitoring method can be configured, either `webapi` or `mqtt`. The other must be set to `enabled: false`.

- **enabled** (Optional - Default: `false`)
    - If set to `true`, Frigate events are collected by polling the web API
- **interval** (Optional - Default: `30`)
    - How frequently to check the Frigate web API for new events, in seconds

```yaml title="Config File Snippet"
frigate:
  webapi:
    enabled: true
    interval: 60
```

#### MQTT

!!! note
    Only one monitoring method can be configured, either `webapi` or `mqtt`. The other must be set to `enabled: false`.

- **enabled** (Optional - Default: `false`)
    - If set to `true`, Frigate events are collected via an MQTT broker
    - Note: This must be the same MQTT broker that Frigate is sending events to
- **server** (Required)
    - IP or hostname of the MQTT server
    - If MQTT monitoring is enabled, this field is required
- **port** (Optiona - Default: `1883`)
    - MQTT service port
- **clientid** (Optional - Default: `frigate-notify`)
    - Client ID of this app used when connecting to MQTT
    - Note: This must be a unique value & cannot be shared with other MQTT clients
- **username** (Optional)
    - MQTT username
    - If username & password are not set, then authentication is disabled
- **password** (Optional)
    - MQTT password
    - Required if `username` is set
- **topic_prefix** (Optional - Default: `frigate`)
    - Optionally change MQTT topic prefix
    - This should match the topic prefix used by Frigate

```yaml title="Config File Snippet"
frigate:
  mqtt: 
    enabled: true
    server: mqtt.your.domain.tld
    port: 1883
    clientid: frigate-notify
    username: mqtt-user
    password: mqtt-pass
    topic_prefix: frigate
```

#### Cameras

- **exclude** (Optional)
    - If desired, provide a list of cameras to ignore
    - Any Frigate events on these cameras will not generate alerts
    - If left empty, this is disabled & all cameras will generate alerts

```yaml title="Config File Snippet"
frigate:
  cameras:
    exclude:
      - test_cam_01
      - test_cam_02
```

### Alerting

!!! note
    Any combination of alerting methods may be enabled, though you'll probably want to enable at least one! 😅

#### General

- **title** (Optional - Default: `Frigate Alert`)
    - Title of alert messages that are generated (Email subject, etc)
- **timeformat** (Optional - Default: `2006-01-02 15:04:05 -0700 MST`)
    - Optionally set a custom date/time format for notifications
    - This utilizes Golang's [reference time](https://go.dev/src/time/format.go) for formatting
    - See [this](https://www.geeksforgeeks.org/time-formatting-in-golang) guide for help
    - Example below uses RFC1123 format

```yaml title="Config File Snippet"
alerts:  
  general:
    title: Frigate Alert
    timeformat: Mon, 02 Jan 2006 15:04:05 MST
```

#### Zones

This config section allows control over whether to generate alerts on all zones, or only specific ones. By default, the app will generate notifications on **all** Frigate events, regardless of whether the event includes zone info.

??? note "A note about how this works"
    
    **With MQTT**, Frigate will send a `new` event when a detection starts. Subsequent changes, like the detected object transitioning from one zone to another, will trigger `update` events. These `update` events will contain a list of current zone(s) that the object is in, as well as a list of all zones that the object has entered during the event.

    In order to reduce the number of notifications generated, this app will only alert on the *first time* the detected object enters a zone.

    For example, let's say you have a camera in your front yard with zones for sidewalk, driveway, and lawn - but only allow notifications for driveway and lawn. During an event someone was detected originally on the sidewalk, then driveway, lawn, and back to driveway. In this case, you should only receive two notifications. Once for the first time the person entered the driveway zone, and a second when they entered the lawn zone. 

    **With Web API event query**, we only query the event from Frigate one time. So currently, only one alert would be sent depending on the detected zones at the time the web API was queried for new events.

- **unzoned** (Optional - Default: `allow`)
    - Controls alerts on events outside a zone
    - By default, events without a zone will generate alerts
    - Set to `drop` to prevent generating alerts from events without a zone
- **allow** (Optional)
    - Specify a list of zones to allow notifications
    - All other zones will be ignored
    - If `unzoned` is set to `allow`, notifications will still be sent on events without any zone info
- **block** (Optional)
    - Specify a list of zones to always ignore
    - This takes precedence over the `allow` list

```yaml title="Config File Snippet"
alerts:
  zones:
    unzoned: allow
    allow:
     - test_zone_01
    block:
     - test_zone_02
```

#### Labels

Similar to [zones](#zones), notifications can be filtered based on labels. By default, the app will generate notifications regardless of any labels received from Frigate. Using this config section, certain labels can be blocked from sending notifications - or an allowlist can be provided to only generate alerts from specified labels.

- **allow** (Optional)
    - Specify a list of labels to allow notifications
    - If set, all other labels will be ignored
    - If not set, all labels will generate notifications
- **block** (Optional)
    - Specify a list of labels to always ignore
    - This takes precedence over the `allow` list

```yaml title="Config File Snippet"
alerts:
  labels:
    allow:
     - person
     - dog
    block:
     - bird
```

#### Discord

- **enabled** (Optional - Default: `false`)
    - Set to `true` to enable alerting via Discord webhooks
- **webhook** (Required)
    - Full URL of the desired Discord webhook to send alerts through
    - Required if this alerting method is enabled
    - Check [Discord's](https://support.discord.com/hc/en-us/articles/228383668-Intro-to-Webhooks) docs for how to create a webhook

```yaml title="Config File Snippet"
alerts:  
  discord:
    enabled: false
    webhook: https://<your-discord-webhook-here>
```

#### Gotify

- **enabled** (Optional - Default: `false`)
    - Set to `true` to enable alerting via Gotify
- **server** (Required)
    - IP or hostname of the target Gotify server
    - Required if this alerting method is enabled
- **token** (Required)
    - App token associated with this app in Gotify
    - Required if this alerting method is enabled
- **ignoressl** (Optional - Default: `false`)
    - Set to `true` to allow self-signed certificates

```yaml title="Config File Snippet"
alerts:  
  gotify:
    enabled: false
    server: gotify.your.domain.tld
    token: ABCDEF
    ignoressl: true
```

#### SMTP

- **enabled** (Optional - Default: `false`)
    - Set to `true` to enable alerting via SMTP
- **server** (Required)
    - IP or hostname of the target SMTP server
    - Required if this alerting method is enabled
- **port** (Required)
    - Port of the target SMTP server
    - Required if this alerting method is enabled
- **tls** (Optional - Default: `false`)
    - Set to `true` if SMTP TLS is required
- **user** (Optional)
    - Add SMTP username for authentication
    - If username & password are not set, then authentication is disabled
- **password** (Optional)
    - Password of SMTP user
    - Required if `user` is set
- **recipient** (Required)
    - Comma-separated list of email recipients
    - Required if this alerting method is enabled

```yaml title="Config File Snippet"
alerts:  
  smtp:
    enabled: false
    server: smtp.your.domain.tld
    port: 587
    tls: true
    user: test_user@your.domain.tld
    password: test_pass
    recipient: nvr_group@your.domain.tld, someone_else@your.domain.tld
```

#### Telegram

In order to use Telegram for alerts, a bot token & chat ID are required.

To obtain a bot token, follow [this](https://core.telegram.org/bots#how-do-i-create-a-bot) doc to message @BotFather.

Once you have a bot token, make sure to initiate a chat message with your bot. Then visit the following URL:

- `https://api.telegram.org/bot<BOT_TOKEN>/getUpdates` 
- Replace `<BOT_TOKEN>` with the API token provided by @BotFather. 

Within the response, locate your message to the bot, then grab the ID under `message > chat > id`. An example response is shown below, where `999999999` is the ID we need to save:

```json
{
  "update_id": 1234567,
  "message": {
    "chat": {
      "id": 999999999,
      "first_name": "Test User",
      "username": "test-username1234",
      "type": "private"
    }
  }
}
```

- **enabled** (Optional - Default: `false`)
    - Set to `true` to enable alerting via Telegram
- **chatid** (Required)
    - Chat ID for the alert recipient
    - Required if this alerting method is enabled
- **token** (Required)

docs/changelog.md
# Changelog

## [v0.3.1](https://github.com/0x2142/frigate-notify/releases/tag/v0.3.0) - Upcoming Release

 - Fix misspelling for Ntfy
     - Currently config will accept either `nfty` or `ntfy`, but in a future release `nfty` will be removed
     - ⚠️ **Note**: If you are using `nfty`, please update your config to the correct spelling: `ntfy`
 - Fix issue where Ntfy action button would use local Frigate URL even if `public_url` was configured
 - Fix camera links for Frigate 0.14.0+
 - Add support for sending HTTP headers with Ntfy notifications
 - Added [config file](https://frigate-notify.0x2142.com/latest/config/file/#general) option to drop events that don't have a snapshot image
 - Move "No snapshot available" message into notification templates

## [v0.3.0](https://github.com/0x2142/frigate-notify/releases/tag/v0.3.0) - Jun 03 2024

 - Reworked & reduced logging, added [debug option](https://frigate-notify.0x2142.com/latest/config/options/)
 - Add ability to define custom [notification templates](https://frigate-notify.0x2142.com/latest/config/templates/#alert-templates)
 - Add support for notifications via [webhook](https://frigate-notify.0x2142.com/latest/config/file/#webhook)
 - Add retry logic to startup check
     - App will now retry up to 5 times, checking once every 30 seconds
     - [Config](https://frigate-notify.0x2142.com/latest/config/file/#server) options now available to modify max attempts & retry interval
 - Enable concurrent notification processing
 - Fix issue where some event providers would fail if no snapshot was available via Frigate

## [v0.2.8](https://github.com/0x2142/frigate-notify/releases/tag/v0.2.8) - May 15 2024

 - Add support for notifications via [Ntfy](https://frigate-notify.0x2142.com/latest/config/file/#ntfy)
 - Add ability to send additional HTTP [headers](https://frigate-notify.0x2142.com/latest/config/file/#frigate) to Frigate
 - Add new `public_url` config item for Frigate
     - This will be used in notification links & should be configured if Frigate is accessible via the internet
 - Add startup check to verify Frigate API is accessible
 - Rework event notifications to be built from templates
 - Fix default interval for querying evens via web API
 - Fix issue where label score is 0% via web API event query

## [v0.2.7](https://github.com/0x2142/frigate-notify/releases/tag/v0.2.7) - May 06 2024

 - Allow changing default MQTT topic prefix via config
 - Added ability to filter notifications based on [labels](https://frigate-notify.0x2142.com/latest/config/file/#labels)
     - New option to allow only notifications with specified labels
     - New option to deny notifications based on labels

## [v0.2.6](https://github.com/0x2142/frigate-notify/releases/tag/v0.2.6) - Apr 01 2024

 - Fixed issue with setting `unzoned: drop` under zone config, where alerts wouldn't be sent if event began outside a zone.
 - Fixed issue with correct timezone getting applied on container image
 - Added `timeformat` option for notifications, which allows custom date/time format
 - Minor updates to logging to ensure all event logs have an event ID attached

## [v0.2.5](https://github.com/0x2142/frigate-notify/releases/tag/v0.2.5) - Mar 29 2024

 - Added support for alerts via [Pushover](https://frigate-notify.0x2142.com/latest/config/file/#pushover)
 - Reduced container image from ~1Gb to ~11Mb

## [v0.2.4](https://github.com/0x2142/frigate-notify/releases/tag/v0.2.4) - Mar 28 2024

 - Added support for alerts via [Telegram](https://frigate-notify.0x2142.com/latest/config/file/#telegram)
 - Updated notification message to now include links to the camera & event clip

## [v0.2.3](https://github.com/0x2142/frigate-notify/releases/tag/v0.2.3) - Mar 14 2024

 - Updated config management
 - Improved config file validation
 - Fixed issue where HTTP requests would fail if only IP & port were specified in config
 - Fixed crash if Frigate has no event info & app couldn't pull snapshots
 - Added ability to filter notifications based on [zones](https://frigate-notify.0x2142.com/latest/config/file/#zones)
     - New config option to only allow notifications based on zone
     - New config option to deny notifications based on zone
     - New config option to allow or drop events outside of a designated zone

## [v0.2.2](https://github.com/0x2142/frigate-notify/releases/tag/v0.2.2) - Nov 29 2023

 - Fix SMTP issue where code was ignoring `tls: false` config flag
 - Fix issue where snapshots would only be sent to the first alerting method, if multiple were enabled
 - Update Debian image used for Docker image

## [v0.2.1](https://github.com/0x2142/frigate-notify/releases/tag/v0.2.1) - Sep 20 2023

 - Added Docker release build for ARM

## [v0.2.0](https://github.com/0x2142/frigate-notify/releases/tag/v0.2.0) - Sep 19 2023

 - Add support for excluding cameras from alerts
 - Added ability to set custom alert titles
 - Added ability to set custom MQTT client ID
 - Added ability to change MQTT port
 - Migrate Discord alerts to use Embeds
 - Improved SMTP & Gotify alerts to include a warning if a snapshot wasn't saved
 - Add GitHub actions for automated release builds
 - Added version number to startup logging

## [v0.1.1](https://github.com/0x2142/frigate-notify/releases/tag/v0.1.1) - Apr 25 2023

 - Add support for MQTT authentication

## [v0.1.0](https://github.com/0x2142/frigate-notify/releases/tag/v0.1.0) - Mar 17 2023

 - Initial release! 🎉


config/config.go
package config

import (
	"encoding/json"
	"fmt"
	"html/template"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/0x2142/frigate-notify/models"
	"github.com/0x2142/frigate-notify/util"
	"github.com/kkyr/fig"
	"github.com/rs/zerolog/log"
)

type Config struct {
	Frigate Frigate `fig:"frigate"`
	Alerts  Alerts  `fig:"alerts"`
	Monitor Monitor `fig:"monitor"`
}

type Frigate struct {
	Server       string              `fig:"server" validate:"required"`
	Insecure     bool                `fig:"ignoressl" default:false`
	PublicURL    string              `fig:"public_url" default:""`
	Headers      []map[string]string `fig:"headers"`
	StartupCheck StartupCheck        `fig:"startup_check"`
	WebAPI       WebAPI              `fig:"webapi"`
	MQTT         MQTT                `fig:"mqtt"`
	Cameras      Cameras             `fig:"cameras"`
	Version      int
}

type StartupCheck struct {
	Attempts int `fig:"attempts" default:5`
	Interval int `fig:"interval" default:30`
}

type WebAPI struct {
	Enabled  bool `fig:"enabled" default:false`
	Interval int  `fig:"interval" default:30`
	TestMode bool `fig:"testmode" default:false`
}

type MQTT struct {
	Enabled     bool   `fig:"enabled" default:false`
	Server      string `fig:"server" default:""`
	Port        int    `fig:"port" default:1883`
	ClientID    string `fig:"clientid" default:"frigate-notify"`
	Username    string `fig:"username" default:""`
	Password    string `fig:"password" default:""`
	TopicPrefix string `fig:"topic_prefix" default:"frigate"`
}

type Cameras struct {
	Exclude []string `fig:"exclude" default:[]`
}

type Alerts struct {
	General  General  `fig:"general"`
	Zones    Zones    `fig:"zones"`
	Labels   Labels   `fig:"labels"`
	Discord  Discord  `fig:"discord"`
	Gotify   Gotify   `fig:"gotify"`
	SMTP     SMTP     `fig:"smtp"`
	Telegram Telegram `fig:"telegram"`
	Pushover Pushover `fig:"pushover"`
	Nfty     Nfty     `fig:"nfty"`
	Ntfy     Ntfy     `fig:"ntfy"`
	Webhook  Webhook  `fig:"webhook"`
}

type General struct {
	Title      string `fig:"title" default:"Frigate Alert"`
	TimeFormat string `fig:"timeformat" default:""`
	NoSnap     string `fig:"nosnap" default:"allow"`
}

type Zones struct {
	Unzoned string   `fig:"unzoned" default:"allow"`
	Allow   []string `fig:"allow" default:[]`
	Block   []string `fig:"block" default:[]`
}

type Labels struct {
	Allow []string `fig:"allow" default:[]`
	Block []string `fig:"block" default:[]`
}

type Discord struct {
	Enabled  bool   `fig:"enabled" default:false`
	Webhook  string `fig:"webhook" default:""`
	Template string `fig:"template" default:""`
}

type Gotify struct {
	Enabled  bool   `fig:"enabled" default:false`
	Server   string `fig:"server" default:""`
	Token    string `fig:"token" default:""`
	Insecure bool   `fig:"ignoressl" default:false`
	Template string `fig:"template" default:""`
}

type SMTP struct {
	Enabled   bool   `fig:"enabled" default:false`
	Server    string `fig:"server" default:""`
	Port      int    `fig:"port" default:25`
	TLS       bool   `fig:"tls" default:false`
	User      string `fig:"user" default:""`
	Password  string `fig:"password" default:""`
	Recipient string `fig:"recipient" default:""`
	Template  string `fig:"template" default:""`
}

type Telegram struct {
	Enabled  bool   `fig:"enabled" default:false`
	ChatID   int64  `fig:"chatid" default:0`
	Token    string `fig:"token" default:""`
	Template string `fig:"template" default:""`
}

type Pushover struct {
	Enabled  bool   `fig:"enabled" default:false`
	Token    string `fig:"token" default:""`
	Userkey  string `fig:"userkey" default:""`
	Devices  string `fig:"devices" default:""`
	Priority int    `fig:"priority" default:0`
	Retry    int    `fig:"retry" default:0`
	Expire   int    `fig:"expire" default:0`
	TTL      int    `fig:"ttl" default:0`
	Template string `fig:"template" default:""`
}

// DEPRECATED: Misspelling of Ntfy
type Nfty struct {
	Enabled  bool                `fig:"enabled" default:false`
	Server   string              `fig:"server" default:""`
	Topic    string              `fig:"topic" default:""`
	Insecure bool                `fig:"ignoressl" default:false`
	Headers  []map[string]string `fig:"headers"`
	Template string              `fig:"template" default:""`
}

type Ntfy struct {
	Enabled  bool                `fig:"enabled" default:false`
	Server   string              `fig:"server" default:""`
	Topic    string              `fig:"topic" default:""`
	Insecure bool                `fig:"ignoressl" default:false`
	Headers  []map[string]string `fig:"headers"`
	Template string              `fig:"template" default:""`
}

type Webhook struct {
	Enabled  bool                   `fig:"enabled" default:false`
	Server   string                 `fig:"server" default:""`
	Insecure bool                   `fig:"ignoressl" default:false`
	Headers  []map[string]string    `fig:"headers"`
	Template map[string]interface{} `fig:"template"`
}

type Monitor struct {
	Enabled  bool   `fig:"enabled" default:false`
	URL      string `fig:"url" default:""`
	Interval int    `fig:"interval" default:60`
	Insecure bool   `fig:"ignoressl" default:false`
}

var ConfigData Config

// loadConfig opens & attempts to parse configuration file
func LoadConfig(configFile string) {
	// Set config file location
	if configFile == "" {
		var ok bool
		configFile, ok = os.LookupEnv("FN_CONFIGFILE")
		if !ok {
			configFile = "./config.yml"
		}
	}

	// Load Config file
	log.Debug().Msgf("Loading config file: %v", configFile)

	err := fig.Load(&ConfigData, fig.File(filepath.Base(configFile)), fig.Dirs(filepath.Dir(configFile)), fig.UseEnv("FN"))
	if err != nil {
		log.Fatal().
			Err(err).
			Msg("Failed to load config file!")
	}
	log.Info().Msg("Config file loaded.")

	// Send config file to validation before completing
	validateConfig()
}

// validateConfig checks config file structure & loads info into associated packages
func validateConfig() {
	var configErrors []string
	var response []byte
	var err error
	log.Debug().Msg("Validating config file...")

	if (ConfigData.Frigate.WebAPI.Enabled && ConfigData.Frigate.MQTT.Enabled) || (!ConfigData.Frigate.WebAPI.Enabled && !ConfigData.Frigate.MQTT.Enabled) {
		configErrors = append(configErrors, "Please configure only one polling method: Frigate Web API or MQTT")
	}

	// Set default web API interval if not specified
	if ConfigData.Frigate.WebAPI.Enabled && ConfigData.Frigate.WebAPI.Interval == 0 {
		ConfigData.Frigate.WebAPI.Interval = 30
	}

	// Warn on test mode being enabled
	if ConfigData.Frigate.WebAPI.Enabled && ConfigData.Frigate.WebAPI.TestMode {
		log.Warn().Msg("~~~~~~~~~~~~~~~~~~~")
		log.Warn().Msg("WARNING: Test Mode is enabled.")
		log.Warn().Msg("This is intended for development only & will only query Frigate for the last event.")
		log.Warn().Msg("Do not enable this in production! App will not accurately check for events.")
		log.Warn().Msg("~~~~~~~~~~~~~~~~~~~")
	}

	// Check if Frigate server URL contains protocol, assume HTTP if not specified
	if !strings.Contains(ConfigData.Frigate.Server, "http://") && !strings.Contains(ConfigData.Frigate.Server, "https://") {
		log.Warn().Msg("No protocol specified on Frigate Server. Assuming http://. If this is incorrect, please adjust the config file.")
		ConfigData.Frigate.Server = fmt.Sprintf("http://%s", ConfigData.Frigate.Server)
	}

	// Test connectivity to Frigate
	log.Debug().Msg("Checking connection to Frigate server...")
	statsAPI := fmt.Sprintf("%s/api/stats", ConfigData.Frigate.Server)
	current_attempt := 1
	if ConfigData.Frigate.StartupCheck.Attempts == 0 {
		ConfigData.Frigate.StartupCheck.Attempts = 5
	}
	if ConfigData.Frigate.StartupCheck.Interval == 0 {
		ConfigData.Frigate.StartupCheck.Interval = 30
	}
	for current_attempt < ConfigData.Frigate.StartupCheck.Attempts {
		response, err = util.HTTPGet(statsAPI, ConfigData.Frigate.Insecure)
		if err != nil {
			log.Warn().
				Err(err).
				Int("attempt", current_attempt).
				Int("max_tries", ConfigData.Frigate.StartupCheck.Attempts).
				Int("interval", ConfigData.Frigate.StartupCheck.Interval).
				Msgf("Cannot reach Frigate server at %v", ConfigData.Frigate.Server)
			time.Sleep(time.Duration(ConfigData.Frigate.StartupCheck.Interval) * time.Second)
			current_attempt += 1
		} else {
			break
		}
	}
	if current_attempt == ConfigData.Frigate.StartupCheck.Attempts {
		log.Fatal().
			Err(err).
			Msgf("Max attempts reached - Cannot reach Frigate server at %v", ConfigData.Frigate.Server)
	}
	var stats models.FrigateStats
	json.Unmarshal([]byte(response), &stats)
	log.Info().Msgf("Successfully connected to %v", ConfigData.Frigate.Server)
	if stats.Service.Version != "" {
		log.Debug().Msgf("Frigate server is running version %v", stats.Service.Version)
		// Save major version number
		ConfigData.Frigate.Version, _ = strconv.Atoi(strings.Split(stats.Service.Version, ".")[1])
	}

	// Check Public / External URL if set
	if ConfigData.Frigate.PublicURL != "" {
		if !strings.Contains(ConfigData.Frigate.PublicURL, "http://") && !strings.Contains(ConfigData.Frigate.PublicURL, "https://") {
			configErrors = append(configErrors, "Public URL must include http:// or https://")
		}
	}

	// Check for camera exclusions
	if len(ConfigData.Frigate.Cameras.Exclude) > 0 {
		log.Debug().Msg("Cameras to exclude from alerting:")
		for _, c := range ConfigData.Frigate.Cameras.Exclude {
			log.Debug().Msgf(" - %v", c)
		}
	}

	// Check MQTT Config
	if ConfigData.Frigate.MQTT.Enabled {
		log.Debug().Msg("MQTT Enabled.")
		if ConfigData.Frigate.MQTT.Server == "" {
			configErrors = append(configErrors, "No MQTT server address specified!")
		}
		if ConfigData.Frigate.MQTT.Username != "" && ConfigData.Frigate.MQTT.Password == "" {
			configErrors = append(configErrors, "MQTT user provided, but no password!")
		}
		if ConfigData.Frigate.MQTT.Port == 0 {
			ConfigData.Frigate.MQTT.Port = 1883
		}
	}

	// Check action on no snapshot available
	if strings.ToLower(ConfigData.Alerts.General.NoSnap) != "allow" && strings.ToLower(ConfigData.Alerts.General.NoSnap) != "drop" {
		configErrors = append(configErrors, "Option for nosnap must be 'allow' or 'drop'")
	} else {
		log.Debug().Msgf("Events without a snapshot: %v", strings.ToLower(ConfigData.Alerts.General.NoSnap))
	}

	// Check Zone filtering config
	if strings.ToLower(ConfigData.Alerts.Zones.Unzoned) != "allow" && strings.ToLower(ConfigData.Alerts.Zones.Unzoned) != "drop" {
		configErrors = append(configErrors, "Option for unzoned events must be 'allow' or 'drop'")
	} else {
		log.Debug().Msgf("Events outside a zone: %v", strings.ToLower(ConfigData.Alerts.Zones.Unzoned))
	}

	if len(ConfigData.Alerts.Zones.Allow) > 0 {
		log.Debug().Msg("Zones to generate alerts for:")
		for _, c := range ConfigData.Alerts.Zones.Allow {
			log.Debug().Msgf(" - %v", c)
		}
	} else {
		log.Debug().Msg("All zones included in alerts")
	}
	if len(ConfigData.Alerts.Zones.Block) > 0 {
		log.Debug().Msg("Zones to exclude from alerting:")
		for _, c := range ConfigData.Alerts.Zones.Block {
			log.Debug().Msgf(" - %v", c)
		}
	} else {
		log.Debug().Msg("No zones excluded")
	}

	// Check Label filtering config
	if len(ConfigData.Alerts.Labels.Allow) > 0 {
		log.Debug().Msg("Labels to generate alerts for:")
		for _, c := range ConfigData.Alerts.Labels.Allow {
			log.Debug().Msgf(" - %v", c)
		}
	} else {
		log.Debug().Msg("All labels included in alerts")
	}
	if len(ConfigData.Alerts.Labels.Block) > 0 {
		log.Debug().Msg("Labels to exclude from alerting:")
		for _, c := range ConfigData.Alerts.Labels.Block {
			log.Debug().Msgf(" - %v", c)
		}
	} else {
		log.Debug().Msg("No labels excluded")
	}

	// Check / Load alerting configuration
	if ConfigData.Alerts.Discord.Enabled {
		log.Debug().Msg("Discord alerting enabled.")
		if ConfigData.Alerts.Discord.Webhook == "" {
			configErrors = append(configErrors, "No Discord webhook specified!")
		}
		// Check template syntax
		if msg := checkTemplate("Discord", ConfigData.Alerts.Discord.Template); msg != "" {
			configErrors = append(configErrors, msg)
		}
	}
	if ConfigData.Alerts.Gotify.Enabled {
		log.Debug().Msg("Gotify alerting enabled.")
		// Check if Gotify server URL contains protocol, assume HTTP if not specified
		if !strings.Contains(ConfigData.Alerts.Gotify.Server, "http://") && !strings.Contains(ConfigData.Alerts.Gotify.Server, "https://") {
			log.Debug().Msg("No protocol specified on Gotify Server. Assuming http://. If this is incorrect, please adjust the config file.")
			ConfigData.Alerts.Gotify.Server = fmt.Sprintf("http://%s", ConfigData.Alerts.Gotify.Server)
		}
		if ConfigData.Alerts.Gotify.Server == "" {
			configErrors = append(configErrors, "No Gotify server specified!")
		}
		if ConfigData.Alerts.Gotify.Token == "" {
			configErrors = append(configErrors, "No Gotify token specified!")
		}
		// Check template syntax
		if msg := checkTemplate("Gotify", ConfigData.Alerts.Gotify.Template); msg != "" {
			configErrors = append(configErrors, msg)
		}
	}
	if ConfigData.Alerts.SMTP.Enabled {
		log.Debug().Msg("SMTP alerting enabled.")
		if ConfigData.Alerts.SMTP.Server == "" {
			configErrors = append(configErrors, "No SMTP server specified!")
		}
		if ConfigData.Alerts.SMTP.Recipient == "" {
			configErrors = append(configErrors, "No SMTP recipients specified!")
		}
		if ConfigData.Alerts.SMTP.User != "" && ConfigData.Alerts.SMTP.Password == "" {
			configErrors = append(configErrors, "SMTP username in config, but no password provided!")
		}
		if ConfigData.Alerts.SMTP.Port == 0 {
			ConfigData.Alerts.SMTP.Port = 25
		}
		// Check template syntax
		if msg := checkTemplate("SMTP", ConfigData.Alerts.SMTP.Template); msg != "" {
			configErrors = append(configErrors, msg)
		}
	}
	if ConfigData.Alerts.Telegram.Enabled {
		log.Debug().Msg("Telegram alerting enabled.")
		if ConfigData.Alerts.Telegram.ChatID == 0 {
			configErrors = append(configErrors, "No Telegram Chat ID specified!")
		}
		if ConfigData.Alerts.Telegram.Token == "" {
			configErrors = append(configErrors, "No Telegram bot token specified!")
		}
		// Check template syntax
		if msg := checkTemplate("Telegram", ConfigData.Alerts.Telegram.Template); msg != "" {
			configErrors = append(configErrors, msg)
		}
	}
	if ConfigData.Alerts.Pushover.Enabled {
		log.Debug().Msg("Pushover alerting enabled.")
		if ConfigData.Alerts.Pushover.Token == "" {
			configErrors = append(configErrors, "No Pushover API token specified!")
		}
		if ConfigData.Alerts.Pushover.Userkey == "" {
			configErrors = append(configErrors, "No Pushover user key specified!")
		}
		if ConfigData.Alerts.Pushover.Priority < -2 || ConfigData.Alerts.Pushover.Priority > 2 {
			configErrors = append(configErrors, "Pushover priority must be between -2 and 2!")
		}
		// Priority 2 is emergency, needs a retry interval & expiration set
		if ConfigData.Alerts.Pushover.Priority == 2 {
			if ConfigData.Alerts.Pushover.Retry == 0 || ConfigData.Alerts.Pushover.Expire == 0 {
				configErrors = append(configErrors, "Pushover retry interval & expiration must be set with priority 2!")
			}
			if ConfigData.Alerts.Pushover.Retry < 30 {
				configErrors = append(configErrors, "Pushover retry cannot be less than 30 seconds!")
			}
		}
		if ConfigData.Alerts.Pushover.TTL < 0 {
			configErrors = append(configErrors, "Pushover TTL cannot be negative!")
		}
		// Check template syntax
		if msg := checkTemplate("Pushover", ConfigData.Alerts.Pushover.Template); msg != "" {
			configErrors = append(configErrors, msg)
		}
	}
	// Deprecation warning
	// TODO: Remove misspelled Ntfy config with v0.4.0 or later
	if ConfigData.Alerts.Nfty.Enabled {
		log.Warn().Msg("Config for 'nfty' will be deprecated due to misspelling. Please update config to 'ntfy'")
		// Copy data to new Ntfy struct
		ConfigData.Alerts.Ntfy.Enabled = ConfigData.Alerts.Nfty.Enabled
		ConfigData.Alerts.Ntfy.Server = ConfigData.Alerts.Nfty.Server
		ConfigData.Alerts.Ntfy.Topic = ConfigData.Alerts.Nfty.Topic
		ConfigData.Alerts.Ntfy.Insecure = ConfigData.Alerts.Nfty.Insecure
		ConfigData.Alerts.Ntfy.Headers = ConfigData.Alerts.Nfty.Headers
		ConfigData.Alerts.Ntfy.Template = ConfigData.Alerts.Nfty.Template
	}
	if ConfigData.Alerts.Ntfy.Enabled {
		log.Debug().Msg("Ntfy alerting enabled.")
		if ConfigData.Alerts.Ntfy.Server == "" {
			configErrors = append(configErrors, "No Ntfy server specified!")
		}
		if ConfigData.Alerts.Ntfy.Topic == "" {
			configErrors = append(configErrors, "No Ntfy topic specified!")
		}
		// Check template syntax
		if msg := checkTemplate("Ntfy", ConfigData.Alerts.Ntfy.Template); msg != "" {
			configErrors = append(configErrors, msg)
		}
	}
	if ConfigData.Alerts.Webhook.Enabled {
		log.Debug().Msg("Webhook alerting enabled.")
		if ConfigData.Alerts.Webhook.Server == "" {
			configErrors = append(configErrors, "No Webhook server specified!")
		}
	}

	// Validate monitoring config
	if ConfigData.Monitor.Enabled {
		log.Debug().Msg("App monitoring enabled.")
		if ConfigData.Monitor.URL == "" {
			configErrors = append(configErrors, "App monitor enabled but no URL specified!")
		}
		if ConfigData.Monitor.Interval == 0 {
			ConfigData.Monitor.Interval = 60
		}
	}

	if len(configErrors) > 0 {
		fmt.Println()
		log.Error().Msg("Config validation failed:")
		for _, msg := range configErrors {
			log.Error().Msgf(" - %v", msg)
		}
		fmt.Println()
		log.Fatal().Msg("Please fix config errors before restarting app.")
	} else {
		log.Info().Msg("Config file validated!")
	}
}

func checkTemplate(provider, alertTemplate string) string {
	var templateError string
	_, err := template.New("").Parse(alertTemplate)
	if err != nil {
		templateError = fmt.Sprintf("Failed to parse %s template: %s", provider, err.Error())
	}
	return templateError
}
